src/machine/lib_machine/mod.rs

```diff
@@ -0,0 +1,632 @@
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+
+use crate::atom_table;
+use crate::heap_iter::{stackful_post_order_iter, NonListElider};
+use crate::machine::machine_indices::VarKey;
+use crate::machine::mock_wam::CompositeOpDir;
+use crate::machine::{
+    F64Offset, F64Ptr, Fixnum, Number, BREAK_FROM_DISPATCH_LOOP_LOC, LIB_QUERY_SUCCESS,
+};
+use crate::parser::ast::{Var, VarPtr};
+use crate::parser::parser::{Parser, Tokens};
+use crate::read::{write_term_to_heap, TermWriteResult};
+
+use dashu::{Integer, Rational};
+use indexmap::IndexMap;
+
+use super::{streams::Stream, Atom, AtomCell, HeapCellValue, HeapCellValueTag, Machine};
+
+#[cfg(test)]
+mod lib_machine_tests;
+
+/// Represents a leaf answer from a query.
+#[derive(Debug, Clone, PartialEq)]
+pub enum LeafAnswer {
+    /// A `true` leaf answer.
+    True,
+    /// A `false` leaf answer.
+    ///
+    /// This means that there are no more answers for the query.
+    False,
+    /// An exception leaf answer.
+    Exception(Term),
+    /// A leaf answer with bindings and residual goals.
+    LeafAnswer {
+        /// The bindings of variables in the query.
+        ///
+        /// Can be empty.
+        bindings: BTreeMap<String, Term>,
+        /// Residual goals.
+        ///
+        /// Can be empty.
+        residual_goals: Vec<Term>,
+    },
```

Comment:
You mention in https://github.com/mthom/scryer-prolog/pull/2582#issuecomment-2526551205 that residual goals are not yet implemented. Would it make sense to remove `residual_goals` here for now and mark the variant as `#[non_exhaustive]` rather than presumably always having an empty residual goal list which might cause someone to believe residual goals are already working.
```suggestion
    /// A leaf answer with bindings and residual goals.
    #[non_exhaustive]
    LeafAnswer {
        /// The bindings of variables in the query.
        ///
        /// Can be empty.
        bindings: BTreeMap<String, Term>,
    },
```

Reply:
Interesting. I was getting pretty close to having residual goals implemented here, building on top of #2527. I was going to have that in this PR, but I decided that stabilizing the interface and unblocking the other fronts was higher priority.

I don't think this would stay unimplemented for too long, certainly not until the next release, and I also can't think of another field for `LeafAnswer::LeafAnswer` apart from `bindings` and `residual_goals`. Do you think it's worth it having this as `#[non_exhaustive]` for safety and then dropping it when (and if) the residual goals are implemented (if they are implemented before the next release and thus still in this major version bump)?

Reply:
@Skgland I personally suggest limiting the scope for this.  There are a lot of things waiting for this, i.e. (selfishly) https://github.com/mthom/scryer-prolog/pull/2465 and there are many interlocking PRs/issues.  

@bakaq I don't know the level of effort required to implement RGs, but if it's significant I suggest limiting the scope here so we don't get bogged down.  Especially if RGs can be positioned as an incremental improvement to this PR.

More small PRs >> less big PRs.  That's my :coin: :coin:

Reply:
I would be fine with this either way. It sounded to me like residual goals would potentially be further out, in which case I think it would have been nicer to make the change, but if residual goals are around the corner I think it's fine the way it is.

Reply:
Does it make sense to distinguish bindings from other goals that are also reported? Potentially yes. On the other hand, why&nbsp;exactly?

Reply:
> On the other hand, why exactly?

Mostly for ease of use. Handling a list of `(=)/2` goals is much clunkier than handling a dictionary, and that's the common case. Is that not a good enough reason? I've thought about that a lot and I think it is worth it.